pkgs/development/mobile/xcodeenv/compose-xcodewrapper.nix

```
@@ -1,19 +1,43 @@
-{ stdenv, lib }:
-{ version ? "11.1"
-, allowHigher ? false
-, xcodeBaseDir ? "/Applications/Xcode.app" }:
+{ lib,
+  stdenv,
+  writeShellScriptBin }:
+{ versions ? [ ] , xcodeBaseDir ? "/Applications/Xcode.app" }:
 
 assert stdenv.isDarwin;
+let
+  xcodebuildPath = "${xcodeBaseDir}/Contents/Developer/usr/bin/xcodebuild";
 
+  xcodebuildWrapper = if builtins.length versions > 0 then
+  writeShellScriptBin "xcodebuild" ''
+    currentVer="$(${xcodebuildPath} -version | awk 'NR==1{print $2}')"
+    wrapperVers=(${lib.concatStringsSep " " versions})
+
+    for ver in "''${wrapperVers[@]}"; do
+      if [[ "$currentVer" == "$ver" ]]; then
+        # here exec replaces the shell without creating a new process
+        # https://www.gnu.org/software/bash/manual/bash.html#index-exec
+        exec "${xcodebuildPath}" "$@"
+      fi
+    done
+
+    echo "The installed Xcode version ($currentVer) does not match any of the allowed versions: ${lib.concatStringsSep ", " versions}"
+    echo "Please update your local Xcode installation to match one of the allowed versions"
+    exit 1
+  ''
+  else
+    writeShellScriptBin "xcodebuild" ''
+      exec "${xcodebuildPath}" "$@"
+  '';
+in
 stdenv.mkDerivation {
-  pname = "xcode-wrapper${lib.optionalString allowHigher "-plus"}";
-  inherit version;
+  pname = "xcode-wrapper-plus";
+  version = if builtins.length versions > 0 then lib.concatStringsSep "," versions else "unwrapped";
   # Fails in sandbox. Use `--option sandbox relaxed` or `--option sandbox false`.
   __noChroot = true;
   buildCommand = ''
     mkdir -p $out/bin
     cd $out/bin
-    ln -s /usr/bin/xcode-select
+    ln -s "${xcodebuildWrapper}/bin/xcode-select"
```

Wouldn't this be simpler?
```suggestion
    ${if versions == [] then ''
    ln -s "${xcodeBaseDir}/Contents/Developer/usr/bin/xcodebuild"
    '' else ''
    ln -s "${xcodebuildWrapper}/bin/xcode-select"
    ''}
```

---

indeed simpler, I've incorporated this change in my latest commit.